### PR TITLE
Added support for adding/editing/removing lights

### PR DESCRIPTION
**Pull request by nebrius**
This implements everything necessary for the lights tab. It also does a handful of under the hood refactors and cleanups. Linting is now unified in the top-level directory, and adds linting to the common folder. It also introduces a new model for Dialogs that's a lot more powerful than the old one.

### CODE REVIEW

File: client/src/components/lib/pageStyles.ts [hunk 1]
```diff
@@ -19,7 +19,7 @@ along with Home Lights.  If not, see <http://www.gnu.org/licenses/>.
```
Comment (allxie) [inline client/src/components/lib/pageStyles.ts:19]: Have you thought about using styledComponents? https://styled-components.com/

Comment (allxie) [reply client/src/components/lib/pageStyles.ts:19]: I think it's compatible with the material-ui lib

File: client/src/components/light/editLightButton.tsx [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+          name="name"
+          description="Friendly name of the light"
+          inputPlaceholder="name"
+          defaultValue={props.light.name}
+        />
+        {reduce(props.light.type, {
+          [LightType.RVL]: () => (
+            <SelectDialogInput
+              name="channel"
+              description="The RVL light channel"
```
Comment (allxie) [inline client/src/components/light/editLightButton.tsx:89]: This is a different description than the one you used in create. (I like this one better)

File: client/src/components/light/lights.tsx [hunk 3]
```diff
@@ -18,8 +18,52 @@ along with Home Lights.  If not, see <http://www.gnu.org/licenses/>.
 */
```
Comment (allxie) [inline client/src/components/light/lights.tsx:19]: Can we rename this component to LightList?

File: client/src/components/light/lights.tsx [hunk 4]
```diff
@@ -18,8 +18,52 @@ along with Home Lights.  If not, see <http://www.gnu.org/licenses/>.
...
+    <div className={classes.container}>
+      <div className={classes.altHeader}>
+        <CreateLightButton
+          createRVLLight={props.createRVLLight}
+          createHueLight={props.createHueLight}
+        />
+      </div>
+      <div className={classes.content}>
+        <div className={classes.innerContent}>
+          <List component="nav" aria-label="main mailbox folders">
```
Comment (allxie) [inline client/src/components/light/lights.tsx:54]: What is main mailbox folders? :-D

File: client/src/components/light/lights.tsx [hunk 5]
```diff
@@ -18,8 +18,52 @@ along with Home Lights.  If not, see <http://www.gnu.org/licenses/>.
...
+            {props.lights.map((light) => (
```
Comment (allxie) [inline client/src/components/light/lights.tsx:55]: This file needs an easter egg that displays "There are four lights" when there are five lights, and "there are five lights" when there are four.

File: client/src/listeners/lights.ts [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+import { Actions } from '../types';
+import {
+  CreateRVLLightRequest,
+  CreateHueLightRequest,
+  Light,
+  LightType
+} from '../common/types';
+import { get, post, put, del } from '../util/api';
+
+listen(Actions.CreateRVLLight, async (name: string, channel: number) => {
```
Comment (allxie) [inline client/src/listeners/lights.ts:30]: Yay!! This file makes so much sense. Reduxology is the best.

File: server/src/db.ts [hunk 7]
```diff
@@ -21,18 +21,36 @@ import { existsSync, mkdirSync } from 'fs';
...
   id INTEGER PRIMARY KEY AUTOINCREMENT,
-  name text NOT NULL UNIQUE
+  name TEXT NOT NULL UNIQUE
+)`;
+
+const LIGHT_SCHEMA = `
+CREATE TABLE "lights" (
+  id INTEGER PRIMARY KEY AUTOINCREMENT,
+  name TEXT NOT NULL UNIQUE,
+  type TEXT NOT NULL,
```
Comment (allxie) [inline server/src/db.ts:52]: I'd love to insert an enum table for this

File: server/src/db.ts [hunk 8]
```diff
@@ -70,3 +91,57 @@ export async function editZone(zone: Zone): Promise<void> {
...
+      await dbRun(`INSERT INTO lights (name, type, channel) values (?, ?, ?)`, [
+        rvlLightRequest.name,
+        LightType.RVL,
+        rvlLightRequest.channel
+      ]);
+      break;
+    }
+    case LightType.Hue: {
+      const hueLightRequest: CreateHueLightRequest = lightRequest as CreateHueLightRequest;
+      await dbRun(`INSERT INTO lights (name, type) values (?, ?)`, [
```
Comment (allxie) [inline server/src/db.ts:116]: Remind me, does this lib protect from sql injection if you parameterize like this?

File: server/src/db.ts [hunk 9]
```diff
@@ -70,3 +91,57 @@ export async function editZone(zone: Zone): Promise<void> {
...
+  }
+}
+
+export async function editLight(light: Light): Promise<void> {
+  switch (light.type) {
+    case LightType.RVL:
+      const rvlLight: RVLLight = light as RVLLight;
+      await dbRun('UPDATE lights SET name = ?, channel = ? WHERE id = ?', [
+        rvlLight.name,
+        rvlLight.channel,
```
Comment (allxie) [inline server/src/db.ts:131]: Should we be doing any validation on the channel?

File: client/src/components/lib/pageStyles.ts (hunk 1 above)
Comment (nebrius) [reply client/src/components/lib/pageStyles.ts:19]: I thought about it, but decided that Material UI's CSS library is good enough. I don't think there's going to be much more styling in the app than what's already here.

File: client/src/components/light/createLightButton.tsx [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+  container: {
+    display: 'flex',
+    'justify-content': 'center',
+    'align-items': 'center'
+  }
+});
+
+export interface CreateLightButtonDispatch {
+  createRVLLight: (name: string, channel: number) => void;
+  createHueLight: (name: string) => void;
```
[the remark this answers is not in the document]
Comment (nebrius) [reply client/src/components/light/createLightButton.tsx:41]: OK, will change

File: client/src/components/light/createLightButton.tsx [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+  props: CreateLightButtonDispatch
+): JSX.Element {
+  const [openDialog, setOpenDialog] = React.useState(false);
+  const [lightType, setLightType] = React.useState(LightType.RVL);
+
+  function handleClose() {
+    setOpenDialog(false);
+  }
+
+  function handleConfirm(values: DialogValue) {
```
[the remark this answers is not in the document]
Comment (nebrius) [reply client/src/components/light/createLightButton.tsx:54]: Weird. I like them and we use the a decent amount at work

File: client/src/components/light/createLightButton.tsx [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with Home Lights.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+import * as React from 'react';
+import { Button } from '@material-ui/core';
+import { Add } from '@material-ui/icons';
```
[the remark this answers is not in the document]
Comment (nebrius) [reply client/src/components/light/createLightButton.tsx:22]: What's the reasoning for this change?

File: client/src/components/light/createLightButton.tsx [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+        onClick={() => setOpenDialog(true)}
+      >
+        <Add />
+      </Button>
+      <Dialog
+        onConfirm={handleConfirm}
+        onCancel={handleClose}
+        onChange={handleChange}
+        open={openDialog}
+        title="Create Zone"
```
[the remark this answers is not in the document]
Comment (nebrius) [reply client/src/components/light/createLightButton.tsx:85]: Nope, will change.

File: client/src/components/light/createLightButton.tsx [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+        description='Enter a descriptive name for the light you wish to add. A
+          light in Home Lights represents a physical light in your home, e.g.
```
[the remark this answers is not in the document]
Comment (nebrius) [reply client/src/components/light/createLightButton.tsx:87]: Sounds good, will change.

File: client/src/components/light/createLightButton.tsx [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+          name="name"
+          description="Friendly name of the light"
+          inputPlaceholder="name"
+        />
+        {reduce(lightType, {
+          [LightType.RVL]: () => (
+            <SelectDialogInput
+              name="channel"
+              description="The RVL light channel"
+              selectValues={Array.from(Array(NUM_RVL_CHANNELS).keys()).map(
```
[the remark this answers is not in the document]
Comment (nebrius) [reply client/src/components/light/createLightButton.tsx:118]: The possible channel values are 0-7 (for now, I'm about to increase it to probably 15), so this allows the user to select which channel they want, and is sent straight to the backend as the `channel` property on the `CreateRVLLightRequest` interface.

File: client/src/components/light/lights.tsx (hunk 3 above)
Comment (nebrius) [reply client/src/components/light/lights.tsx:19]: It does more than show a list, since it also has the create light button in it. This file is directly analogous to the zone/zones.tsx file, and I followed the naming scheme there.

File: client/src/components/light/lights.tsx (hunk 4 above)
Comment (nebrius) [reply client/src/components/light/lights.tsx:54]: Something that was copy pasted 😂

File: client/src/listeners/lights.ts (hunk 6 above)
Comment (nebrius) [reply client/src/listeners/lights.ts:30]: 😁

File: server/src/db.ts (hunk 7 above)
Comment (nebrius) [reply server/src/db.ts:52]: My first though too, but SQLite doesn't support enums. I'm not worried about it though, since it's backed by an enum in TypeScript, and so will have the same effect.

File: server/src/db.ts (hunk 8 above)
Comment (nebrius) [reply server/src/db.ts:116]: Yes it does, see https://github.com/mapbox/node-sqlite3/wiki/API#databaserunsql-param--callback

File: server/src/db.ts (hunk 9 above)
Comment (nebrius) [reply server/src/db.ts:131]: Yeah sure, will do.

File: client/src/components/light/createLightButton.tsx (hunk 14 above)
[the remark this answers is not in the document]
Comment (nebrius) [reply client/src/components/light/createLightButton.tsx:87]: Oh, so looking at the code, I realized something subtle about all this. In the case of the Zones, we now use the InputDialog wrapper around the more generic Dialog infrastructure I created here. This obscures a detail though: there's a description for the dialog _and_ for the text input separately, but the wrapper treats it like a single description. So contextually the two are different, but I treated them sorta like the same here. Needs a larger refactor I think

[the remark this answers is not in the document]
Comment (nebrius) [reply client/src/components/light/createLightButton.tsx:87]: I also filed https://github.com/rvl-system/home-lights/issues/26 to add UI validation.

File: client/src/components/light/editLightButton.tsx (hunk 2 above)
Comment (nebrius) [reply client/src/components/light/editLightButton.tsx:89]: Could you be looking at a difference description perhaps? This is identical to the one in create.

File: client/src/components/light/createLightButton.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (allxie) [reply client/src/components/light/createLightButton.tsx:41]: (changed in different branch)

File: client/src/components/light/deleteLightButton.tsx [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+
+      <Dialog
+        onConfirm={() => {
+          handleDeleteClose();
+          props.deleteLight(props.light.id);
+        }}
+        onCancel={handleDeleteClose}
+        open={deleteDialogOpen}
+        title="Delete Light"
+        description={`Are you sure you want to delete light ${props.light.name}?`}
```
Comment (allxie) [inline client/src/components/light/deleteLightButton.tsx:65]:
```suggestion
        description={`Are you sure you want to delete ${props.light.name}?`}
```

File: client/src/components/zone/createZoneButton.tsx [hunk 17]
```diff
@@ -62,10 +62,8 @@ export const CreateZoneButton: FunctionComponent<CreateZoneButtonDispatch> = (
         onCancel={handleClose}
         open={openDialog}
         title="Create Zone"
-        description='Enter a descriptive name for the zone you wish to add. A zone in
-          Home Lights represents a physical area in your home, e.g.
-          "kitchen," "guest bedzone", "left side bed nightstand" etc. The zone name
-          must not already be in use.'
+        description='Enter a descriptive name for the zone you wish to add, e.g.
+          "kitchen," "guest bedzone", "left side bed nightstand" etc.'
```
Comment (allxie) [inline client/src/components/zone/createZoneButton.tsx:66]:
```suggestion
        description='Room name of the zone you wish to add'
```

File: client/src/components/zone/editZoneButton.tsx [hunk 18]
```diff
@@ -53,18 +61,12 @@ export const EditZoneButton: FunctionComponent<
...
-          handleEditClose();
-          props.editZone({
-            ...props.zone,
-            name
-          });
-        }}
+        onConfirm={handleEditConfirm}
         onCancel={handleEditClose}
         open={editDialogOpen}
         title="Edit Zone"
```
Comment (allxie) [inline client/src/components/zone/editZoneButton.tsx:72]: Would you please shorten this to match the description I put in create zone?

**Review by allxie — APPROVED** [review]
